app/src/main/java/com/dalvik/dimadenyslessons/review/Review.kt

```diff
@@ -0,0 +1,177 @@
+                updateAdState(adUnitId, Result.Error(Throwable("Failed to load")))
+            }
+
+            override fun onAdDisplayFailed(p0: MaxAd, p1: MaxError) {
+
+            }
+        }
+
+        private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
+        private val mutex = Mutex()
+
+        private val interstitials = adUnits.map {
+            MaxInterstitialAdWrapper(
+                adUnitId = it,
+                maxAd = MaxInterstitialAd(it, context).apply {
+                    setListener(maxInterstitialAdListener)
+                }
+            )
+        }
+
+        init {
+            startWaterfallLoading(loadTimeout)
+        }
+
+        private fun startWaterfallLoading(timeout: Long){
+            scope.launch {
+                val states = interstitials.map { it.state }
+                val allAdsLoading = states.all { it is Result.Loading }
+
+                if (!allAdsLoading){
+                    for (interstitial in interstitials){
+                        when (interstitial.state){
+                            is Result.Initial, is Result.Error -> {
+                                mutex.withLock {
+                                    interstitial.state = Result.Loading
+                                }
+
+                                withTimeoutOrNull(timeout){
+                                    loadAd(interstitial)
+                                }
+
+                                if (interstitial.state is Result.Success){
+                                    break
+                                }
+                            }
+
+                            is Result.Loading -> {
+                                delay(timeout)
+
+                                if (interstitial.state is Result.Success){
+                                    break
+                                }
+                            }
+
+                            is Result.Success -> {
+
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private suspend fun loadAd(maxInterstitialAdWrapper: MaxInterstitialAdWrapper): Unit = suspendCancellableCoroutine { continuation ->
+            maxInterstitialAdWrapper.maxAd.loadAd()
+        }
+
+        private fun updateAdState(adUnitId: String, state: Result<MaxAd>) {
+            scope.launch {
+                mutex.withLock {
+                    interstitials.firstOrNull{ it.adUnitId == adUnitId}?.let { it.state = state }
+                }
+            }
+        }
+
+        override fun showInterstitial(activity: Activity): Boolean {
+            val loadedAd = interstitials.map { it.maxAd }.firstOrNull { maxAd -> maxAd.isReady }
+
+            return loadedAd?.let {
+                it.showAd(activity)
+                true
+            } ?: run {
+                startWaterfallLoading(loadTimeout)
```

Comment:
does it have to retry automatically even if the result won't be shown automatically? It only starts re-loading again, but does nothing with the result of loading, so this retry functionality seems like a side effect to me. I'd better exposed another method `reloadContent()` and called it every time when showInterstitial() returns false.